Azure Functions Cli

ahmelsayed opened the pull request:
```bash
> npm i -g azurefunctions
```

##### Local -> Azure:
```bash
> mkdir functions
> cd functions
> func init
> func new # launches yoeman generator (assume functionName=HttpCSharp)
>  func run HttpCSharp -f file.json
> git commit -am "First function"
> func new functionapp MyNewAwesomeFunctionApp -l WestUS
> func list functionapps # displays the git url for the function app
> func user # lets me reset publishing user.
>  git push https://MyNewAwesomeFunctionApp.scm.azurewebsites.net/ master
> func open MyNewAwesomeFunctionApp # launches default web browser with deep link to portal
```

##### Azure -> Local:
```bash
> func login
> func list functionapps # displays a list of function apps
> func switch-tenants # displays a list of tenants
> func switch-tenants <tenantId> 
> func list functionapps
> git clone https://MyNewAwesomeFunctionApp.scm.azurewebsites.net
> cd MyNewAwesomeFunctionApp
> func fetch secrets MyNewAwesomeFunctionApp
> func run FunctionName # then you go in the portal and switch to local
```

```bash
> func help
Azure Functions CLI 0.1
Usage: func [verb] [Options]

   config          get and set global cli config options
   fetch           fetches function app secrets
   login           Clears login cache and prompts for a new login
   logout          Clears login cache
   open            Launch default browser with link to the function app in https://portal.azure.com
   switch-tenants  List and switch current tenant for the Cli
   init            Creates .gitignore, and host.json. Runs git init .
   set             Not yet Implemented
   new             Handle creating a new function or function app
   run             Run the specified function locally
   user            Helps manage publishing username
   web             Launches a Functions server endpoint locally
   list            Lists function apps in current tenant. See switch-tenant command


Tip: run func init to get started.
```

```bash
>  azf help new
Azure Functions CLI 0.1
Usage: func new [function/functionApp] <functionAppName> [Options]

   <newOption>          [Function/FunctionApp/StorageAccount/Secret]
   <functionAppName>    (Required)
   -s/--subscription    Subscription to create function app in
   -l/--location        Geographical location for your function app

Tip: run func init to get started.
```

```bash
>  func help run
Azure Functions CLI 0.1
Usage: func run <functionName> [Options]

   <functionName>    (Required)
   -t/--timeout      Time to wait until Functions Server is ready in Seconds
   -c/--content      In line content to use
   -f/--file         File name to use as content
   -d/--debug        Attach a debugger to the host process before running the function.

Tip: run func init to get started.
```


```bash
>  func help list
Azure Functions CLI 0.1
Usage: func list FunctionApps [Options]

Usage: func list Secrets [Options]

   -a/--show    Display the secret value

Usage: func list StorageAccounts [Options]

Usage: func list Tenants [Options]

Tip: run func init to get started.
```


```bash
>  func help user
Azure Functions CLI 0.1
Usage: func user <userName> [Options]

   <userName>    (Required)

Tip: run func init to get started.
```

```bash
>  func help config
Azure Functions CLI 0.1
Usage: func config <name> <value> [Options]

   <name>     (Required)
   <value>    (Required)
```

dnfclas commented:
Hi __@ahmelsayed__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

awolf commented:
Hi @ahmelsayed, as a .Net developer doing Node development, this is the interface I would be most comfortable with. Especially since I use a Mac and Atom for my development environment.

The command are logical, well thought out and not surprising. That's always a plus.

mathewc commented on src/WebJobs.Script.ConsoleHost/App.config:
```diff
@@ -0,0 +1,42 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Can you name this project (and related projects) WebJobs.Script.CLI? That leads to less confusion with the existing 2 hosts we have, which are about hosting functions, not CLI stuff.

mathewc commented on src/WebJobs.Script.ConsoleHost/Arm/Constants.cs:
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
```
nit: remove extra spaces here and in other files

mathewc commented on src/WebJobs.Script.ConsoleHost/Common/Verify.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public static void IsPfxCert(string path)
+        {
+
```
Either implement or remove?

mathewc commented on src/WebJobs.Script.ConsoleHost/Common/ConsoleTracer.cs:
```diff
@@ -0,0 +1,21 @@
+
+        public override void Trace(TraceEvent traceEvent)
+        {
+            if (traceEvent.Source == Constants.CliTracingSource)
```
nit: missing braces

mathewc commented on src/WebJobs.Script.ConsoleHost/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,40 @@
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
+[assembly: InternalsVisibleTo("WebJobs.Script.ConsoleHost.Tests")]
```
You can add the following line to disable parallelism:

```csharp
[assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly, DisableTestParallelization = true)]
```

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/FunctionAppBaseVerb.cs:
```diff
@@ -0,0 +1,18 @@
+
+        public void FunctionAppNotFound()
+        {
+            TraceInfo($"Can't Find Function App Named {FunctionAppName} in tenant");
```
casing nit: Can't find Function App named ...

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/HelpVerb.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public override Task Run()
+        {
+            TraceInfo("azf (Azure Functions CLI) 0.1");
```
Are you going to rename this to "func" rather than "azf" before checking in?

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/ListVerb.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace WebJobs.Script.ConsoleHost.Verbs
+{
+    [Verb(HelpText = "Lists function apps in current tenant. See `switch-tenant` command")]
```
nit: make casing consistent across all output. You use "Function App" other places.

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/NewVerb.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace WebJobs.Script.ConsoleHost.Verbs
+{
+    [Verb(HelpText = "Handle creating a new Function or FunctionApp", Usage = "[function/functionApp] <functionAppName>")]
```
and here you're using all one word FunctionApp :)

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/OpenVerb.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace WebJobs.Script.ConsoleHost.Verbs
+{
+    [Verb(HelpText = "Launche default browser with link to functionApp in https://portal.azure.com" )]
```
nit: spelling Launche

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/RunVerb.cs:
```diff
@@ -0,0 +1,57 @@
+        [Option(0)]
+        public string FunctionName { get; set; }
+
+        [Option('t', "timeout", DefaultValue = 10, HelpText = "Time to wait until Functions Server is read in Seconds")]
```
did you mean "ready" not "read"?

mathewc commented on src/WebJobs.Script.ConsoleHost/Verbs/SwitchTenantsVerb.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace WebJobs.Script.ConsoleHost.Verbs
+{
+    [Verb("switch-tenants", HelpText = "List and switche current tenant for the Cli", Usage = "<tenantId>")]
```
nit: spelling "switch"

mathewc commented on src/WebJobs.Script.ConsoleHost/WebJobs.Script.ConsoleHost.csproj:
```diff
@@ -0,0 +1,253 @@
+    <FileAlignment>512</FileAlignment>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Add code analysis to all new projects. See other projects for reference. I.e.:

```xml
    <RunCodeAnalysis>true</RunCodeAnalysis>
    <CodeAnalysisRuleSet>..\src.ruleset</CodeAnalysisRuleSet>
    <CodeAnalysisCulture>en-US</CodeAnalysisCulture>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <StyleCopTreatErrorsAsWarnings>false</StyleCopTreatErrorsAsWarnings>
```

mathewc commented on src/WebJobs.Script.ConsoleHost/WebJobs.Script.ConsoleHost.v2.ncrunchproject:
```diff
@@ -0,0 +1,29 @@
+  <PreloadReferencedAssemblies>false</PreloadReferencedAssemblies>
+  <AllowDynamicCodeContractChecking>true</AllowDynamicCodeContractChecking>
+  <AllowStaticCodeContractChecking>false</AllowStaticCodeContractChecking>
+  <AllowCodeAnalysis>false</AllowCodeAnalysis>
```
I won't pretend to know what an "ncrunch" is. Do we need this? Can't we just have a regular 'ol project? :)

ahmelsayed replied:
ncrunch projects are cruncher than normal project.

mathewc commented on src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs:
```diff
@@ -17,9 +18,15 @@ internal static void Initialize(ContainerBuilder builder, WebHostSettings settin
+                TraceWriter = settings.TraceWriter,
             };
 
+            if (settings.IsSelfHost && settings.TraceWriter != null)
```
Can you add a comment here? Is your intent to disable all logging? Perhaps you don't want FileLoggingEnabled above as well?

mathewc commented on src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs:
```diff
@@ -36,6 +43,13 @@ internal static void Initialize(ContainerBuilder builder, WebHostSettings settin
             }
+            else if (settings.IsSelfHost)
+            {
+                // TODO: persist
```
Address before checking in

mathewc commented on src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs:
```diff
@@ -36,6 +43,13 @@ internal static void Initialize(ContainerBuilder builder, WebHostSettings settin
+            else if (settings.IsSelfHost)
+            {
+                // TODO: persist
+                scriptHostConfig.HostConfig.HostId = Guid.NewGuid().ToString().Replace("-", string.Empty).ToLowerInvariant();
```
We use guid.ToString("N") in WebJobs/Functions for host ids

mathewc commented on src/WebJobs.Script.WebHost/Controllers/AdminController.cs:
```diff
@@ -30,7 +30,7 @@ public AdminController(WebScriptHostManager scriptHostManager)
         [HttpPost]
         [Route("admin/functions/{name}")]
-        public HttpResponseMessage Invoke(string name, [FromBody] FunctionInvocation invocation)
+        public async Task<HttpResponseMessage> Invoke(string name, [FromBody] FunctionInvocation invocation, bool block = false)
```
How about adding a "WaitForCompletion" property to FunctionInvocation? That was the intent of that type - to contain all the various invocation flags/data.

mathewc commented on src/WebJobs.Script.WebHost/Controllers/AdminController.cs:
```diff
@@ -48,7 +48,12 @@ public HttpResponseMessage Invoke(string name, [FromBody] FunctionInvocation inv
+
+            if (block)
+            {
+                await runTask;
```
If we're waiting for completion, we don't want to return an Accepted status code, but rather an 200 or 500 based on whether the invocation completed successfully.

mathewc commented on src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs:
```diff
@@ -83,8 +83,10 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
+
                 // Authorize
-                if (authorizationLevel < httpFunctionMetadata.AuthLevel)
+                if (authorizationLevel < httpFunctionMetadata.AuthLevel && !settings.IsSelfHost)
```
Is your intent to bypass auth for all local requests? Then perhaps you should set authorizationLevel to Admin above when we determine it. As it stands now, you'll still be doing auth checks on WebHooks

mathewc commented on test/WebJobs.Script.ConsoleHost.Tests/ArmTests/ArmManagerTests.cs:
```diff
@@ -0,0 +1,18 @@
+        [Fact]
+        public async Task GetFunctionAppsTest()
+        {
+
```
This doesn't count as a test :)

mathewc commented on test/WebJobs.Script.ConsoleHost.Tests/WebJobs.Script.ConsoleHost.Tests.csproj:
```diff
@@ -0,0 +1,153 @@
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>WebJobs.Script.ConsoleHost.Tests</RootNamespace>
+    <AssemblyName>WebJobs.Script.ConsoleHost.Tests</AssemblyName>
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
```
Add the standard code analysis blocks here as well (as in other test projects)

mathewc commented:
You'll want to add entries to WebJobs.Script.proj so this is building on the CI machine and tests are running. @fabiocav can also help you make sure artifacts are being generated as part of our AppVeyor build (we'll want to start including these in our milestone release drops, e.g. [here](https://github.com/Azure/azure-webjobs-sdk-script/releases/tag/v1.0.0-beta1-10307)).

mathewc commented:
Also you'll want to start a [wiki page](https://github.com/Azure/azure-webjobs-sdk-script/wiki) on how to use the CLI :) At least getting people started.

fabiocav commented on src/WebJobs.Script.ConsoleHost/Arm/Models/ArmArrayWrapper.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace WebJobs.Script.ConsoleHost.Arm.Models
+{
+    public class ArmArrayWrapper<T>
```
I'm assuming this class exists to aid serialization, is that correct? Would probably be a good idea to document this class.

fabiocav commented on src/WebJobs.Script.ConsoleHost/Common/FileTracer.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public class FileTracer : TraceWriter
+    {
+        private Collection<TraceEvent> _traces = new Collection<TraceEvent>();
```
Unused?

fabiocav commented on src/WebJobs.Script.ConsoleHost/Common/Functish.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace WebJobs.Script.ConsoleHost.Common
+{
+    public enum Functish
```
Functish? Should we come up with a better name? Is this some yeoman lingo?

ahmelsayed replied:
yeah, we should come up with a better name :). I just needed something that could represent that enum. It's a LibGit2 lingo. They have what they call a committish, that can be commit id, branch, tag, etc. If you have a better name I'm all ears :) otherwise I'll think of something

fabiocav commented on src/WebJobs.Script.ConsoleHost/Verbs/KillVerb.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace WebJobs.Script.ConsoleHost.Verbs
+{
+    [Verb(HelpText = "Kills running server instances of azf")]
```
The string "azf" (or whatever the CLI ends up being called) should probably be moved to your constants.

fabiocav commented on src/WebJobs.Script.ConsoleHost/Verbs/NewVerb.cs:
```diff
@@ -0,0 +1,60 @@
+                var subscriptions = await _armManager.GetSubscriptions();
+                if (string.IsNullOrEmpty(Subscription) && subscriptions.Count() != 1)
+                {
+                    TraceInfo("Can't determin subscription Id, please add -s/--subscription <SubId>");
```
nit: Spelling "determin"

fabiocav commented on src/WebJobs.Script.ConsoleHost/Verbs/WebVerb.cs:
```diff
@@ -0,0 +1,95 @@
+        [Option('p', "port", DefaultValue = 6061, HelpText = "Local port to listen on")]
+        public int Port { get; set; }
+
+        [Option('c', "cert", HelpText = "Path for the cert to use. If not supecified, will auto-generate a cert")]
```
Spelling "supecified"

fabiocav commented on test/WebJobs.Script.Tests/WebJobs.Script.Tests.v2.ncrunchproject:
```diff
@@ -0,0 +1,30 @@
+﻿<ProjectConfiguration>
```
Should probably add nchrunchproject files to .gitignore

fabiocav commented:
Took a high level pass. Not a lot of feedback on the actual implementation, but here are a couple of points in addition to the comments I've left:

- bit of a nit (and often source of much debate), but we may want to add the "Async" suffix to Task returning methods to follow the convention used in the project.
- There are several spots where we're missing brackets on if statements and others, enabling the checks Mathew mentioned should point those out.

ahmelsayed commented:
@mathewc @fabiocav I addressed all the comments, except for renaming the functions that return Tasks to end in Async. Do I really have to do that :) If it's a style issue for the project, I've no problem renaming them, just encourage you to re-question the style. I know that it's a debatable thing, but the way I look at it, if there is only 1 option for an API (that is async, no sync version) and it's clear from the definition that it returns task, it's it implicit and clear that it's an async function? Adding a prefix to the name seems redundant.

fabiocav commented:
Oh, need to play with it now since I'm sure 0326a20 made is so pretty 

Now all we need is some ASCII art! :)

ahmelsayed commented:
:) not all commands have colors now, but yes, I was hoping that adding some color would make the output and help more readable, especially in error scenarios or scenarios that require user action.

christopheranderson commented:
0.0.10 isn't installing on npm - 

> npm install -g azurefunctions
npm ERR! cb() never called!

0.0.9 installs successfully

ahmelsayed commented:
you most probably have a server already running in the background. try running `func kill` to see what's running and `func kill -a` if you want to kill all of them.

christopheranderson commented:
Still a failure:

![image](https://cloud.githubusercontent.com/assets/2252976/17471993/fe8ce458-5cfc-11e6-8a7b-49d0369a0b63.png)

ahmelsayed commented:
can you try `npm cache clean`?

christopheranderson commented:
Still having issues running this on my machine. I'm getting an error thrown from Ln23 of ConsoleApp.cs. This repros on 0.0.9 as well.

```
public static void Run<T>(string[] args, IDependencyResolver dependencyResolver = null)
        {
            Task.Run(() => RunAsync<T>(args, dependencyResolver)).Wait();
        }
```

```
Exception thrown: 'Microsoft.WindowsAzure.Storage.StorageException' in Microsoft.WindowsAzure.Storage.dll
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.IO.FileNotFoundException' in mscorlib.dll
Exception thrown: 'Microsoft.WindowsAzure.Storage.StorageException' in Microsoft.WindowsAzure.Storage.dll
The thread 0xa20 has exited with code 0 (0x0).
Exception thrown: 'Microsoft.WindowsAzure.Storage.StorageException' in Microsoft.WindowsAzure.Storage.dll
Exception thrown: 'System.InvalidOperationException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
The program '[14168] Func.vshost.exe' has exited with code -1 (0xffffffff).
```

Looks like Storage issues? I tried setting environment variables/etc.

I am also still getting install issues with 0.0.10, but I manually grabbed the tarball, unzipped it, and placed it in my global npm. It runs fine, except for `web`/`run`.

```
> func run .\httptest\
Response Status Code: ServiceUnavailable
```

It's quite hard to tell if I'm doing something wrong or if it's just not functional. Attaching a debugger only kept me going through a loop where I was getting the above exceptions being thrown.

Ideas for next steps?

ahmelsayed commented:
are you running `Func.vshost.exe` or `func.exe`? also it might be a legitimate issue. Maybe I have some storage dll globally in my gac or something and it's not included in the package.

ahmelsayed commented:
Also the web command is the only command that actually loads the host, so it's the one most likely to fail because of a missing dependency.

christopheranderson commented:
I'm not running func.exe/func.vshost.exe directly in either case. In one, I attached the debugger to the source in your fork to find where the exception was being thrown. In the other, it was just the npm package that gets installed from npm registry. I tried attaching to the host to see if I can find in more detail what was causing the host to not start, but symbols weren't loading for the script.host

christopheranderson commented:
Also, I realized I didn't give verbose logs for the npm install issue:

```
C:\Users\chrande> npm cache clean
C:\Users\chrande> npm install -g azurefunctions --verbose
npm info it worked if it ends with ok
npm verb cli [ 'C:\\Program Files\\nodejs\\node.exe',
npm verb cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
npm verb cli   'install',
npm verb cli   '-g',
npm verb cli   'azurefunctions',
npm verb cli   '--verbose' ]
npm info using npm@3.7.3
npm info using node@v5.9.1
npm verb request uri https://registry.npmjs.org/azurefunctions
npm verb request no auth needed
npm info attempt registry request try #1 at 10:40:23 AM
npm verb request id f95c9ebcb62b4b9a
npm http request GET https://registry.npmjs.org/azurefunctions
npm http 200 https://registry.npmjs.org/azurefunctions
npm verb headers { server: 'CouchDB/1.5.0 (Erlang OTP/R16B03)',
npm verb headers   etag: '"27SBPYMH9HQDLMHKR46X8VOFY"',
npm verb headers   'content-type': 'application/json',
npm verb headers   'content-encoding': 'gzip',
npm verb headers   'cache-control': 'max-age=300',
npm verb headers   'content-length': '1796',
npm verb headers   'accept-ranges': 'bytes',
npm verb headers   date: 'Tue, 09 Aug 2016 17:40:27 GMT',
npm verb headers   via: '1.1 varnish',
npm verb headers   age: '0',
npm verb headers   connection: 'keep-alive',
npm verb headers   'x-served-by': 'cache-sjc3622-SJC',
npm verb headers   'x-cache': 'MISS',
npm verb headers   'x-cache-hits': '0',
npm verb headers   'x-timer': 'S1470764427.365202,VS0,VE142',
npm verb headers   vary: 'Accept-Encoding' }
npm verb get saving azurefunctions to C:\Users\chrande\AppData\Roaming\npm-cache\registry.npmjs.org\azurefunctions\.cach
e.json
npm verb correctMkdir C:\Users\chrande\AppData\Roaming\npm-cache correctMkdir not in flight; initializing
npm verb makeDirectory C:\Users\chrande\AppData\Roaming\npm-cache creation not in flight; initializing
npm verb makeCacheDir UID & GID are irrelevant on win32
npm verb cache add spec azurefunctions
npm verb addNamed "latest" is being treated as a dist-tag for azurefunctions
npm info addNameTag [ 'azurefunctions', 'latest' ]
npm verb addNameTag registry:https://registry.npmjs.org/azurefunctions not in flight; fetching
npm verb get https://registry.npmjs.org/azurefunctions not expired, no request
npm verb addNamed "0.0.10" is a plain semver version for azurefunctions
npm verb addRemoteTarball https://registry.npmjs.org/azurefunctions/-/azurefunctions-0.0.10.tgz not in flight; adding
npm verb addRemoteTarball [ 'https://registry.npmjs.org/azurefunctions/-/azurefunctions-0.0.10.tgz',
npm verb addRemoteTarball   '51f6bce1e29f4d037421255b9b370822e75605a3' ]
npm info retry fetch attempt 1 at 10:40:24 AM
npm info attempt registry request try #1 at 10:40:24 AM
npm http fetch GET https://registry.npmjs.org/azurefunctions/-/azurefunctions-0.0.10.tgz
npm http fetch 200 https://registry.npmjs.org/azurefunctions/-/azurefunctions-0.0.10.tgz
npm verb addTmpTarball C:\Users\chrande\AppData\Local\Temp\npm-4228-e4681199\registry.npmjs.org\azurefunctions\-\azurefu
nctions-0.0.10.tgz not in flight; adding
npm verb addTmpTarball already have metadata; skipping unpack for azurefunctions@0.0.10
npm verb correctMkdir C:\Users\chrande\AppData\Roaming\npm-cache correctMkdir not in flight; initializing
npm verb afterAdd C:\Users\chrande\AppData\Roaming\npm-cache\azurefunctions\0.0.10\package\package.json not in flight; w
riting
npm verb correctMkdir C:\Users\chrande\AppData\Roaming\npm-cache correctMkdir not in flight; initializing
npm verb afterAdd C:\Users\chrande\AppData\Roaming\npm-cache\azurefunctions\0.0.10\package\package.json written
npm ERR! cb() never called!
C:\Users\chrande>
```

It doesn't look like it's happening in your code, but I've not having this happen on any other package I've tried lately.

christopheranderson commented:
I found the issue with run/web not working - I had a environment variable on all of my machines from a previous experiment which was being picked up and was broken. :) Works nicely now. Having a blast!

christopheranderson commented:
That probably does highlight some of the issues with our current model for running the host in background - we need a better debugging experience for host issues.
